src/core/Oscillator.cpp

```diff
@@ -88,6 +118,206 @@ void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
+		}
+		max = fmax(max, m_generatedWaveTable[i]);
+	}
+	for (int i = 0; i < WAVETABLE_LENGTH; i++)
+	{
+		m_generatedWaveTable[i] /= max;
+	}
+}
+
+
+
+
+//expects sample in sample buffer
+void Oscillator::generateFromFFT(int bands, float threshold)
+{
+	float max = 0;
+	
+	//set unrequired bands to zero 
+	for (int i = bands; i < WAVETABLE_LENGTH * 2 + 1 - bands; i++)
+	{
+		m_specBuf[i][0] = 0.0f;
+		m_specBuf[i][1] = 0.0f;
+	}
+	for (int i = 0; i < WAVETABLE_LENGTH * 2 + 1; ++i)
+	{
+		if (m_specBuf[i][0] * m_specBuf[i][0] + m_specBuf[i][1] * m_specBuf[i][1] < threshold)
+		{
+			m_specBuf[i][0] = 0.0f;
+			m_specBuf[i][1] = 0.0f;
+		}
+	}
+	//ifft
+	fftwf_execute(m_ifftPlan);
+	//normalise
+	for (int i = 0; i < WAVETABLE_LENGTH; ++i)
+	{
+		max = fmax(max, m_sampleBuffer[i]);
+	}
+	//copy to generateWaveTable
+	for (int i = 0; i < WAVETABLE_LENGTH; ++i)
+	{
+		m_generatedWaveTable[i] = m_sampleBuffer[i] / max;
+	}
+}
+
+
+int Oscillator::waveTableBandFromFreq(float freq)
+{
+	int i;
+	for (i = 0; i < s_waveTablesPerWaveformCount; ++i)
+	{
+		if (s_waveTableBandFreqs[i] > freq) return i;
+	}
+	return i;
+}
+
+sample_t ***Oscillator::s_waveTables = 0;
+float* Oscillator::s_waveTableBandFreqs = 0;
+int Oscillator::s_waveTablesPerWaveformCount = 0;
+
+
+
+void Oscillator::allocWaveTables()
+{
+	s_waveTables = new sample_t**[WaveShapes::NumWaveShapes];
+	memset(s_waveTables, 0, sizeof(sample_t) * WaveShapes::NumWaveShapes);
+	for (int i = 0; i < WaveShapes::NumWaveShapes; ++i)
+	{
+		s_waveTables[i] = new sample_t*[s_waveTablesPerWaveformCount +1];
+		memset(s_waveTables[i], 0, sizeof(sample_t) * s_waveTablesPerWaveformCount);
+	}
+}
+
+void Oscillator::createFFtPlans()
+{
+	m_specBuf = ( fftwf_complex * ) fftwf_malloc( ( WAVETABLE_LENGTH * 2 + 1 ) * sizeof( fftwf_complex ) );
+	m_fftPlan = fftwf_plan_dft_r2c_1d( WAVETABLE_LENGTH , m_sampleBuffer, m_specBuf, FFTW_MEASURE );
+	m_ifftPlan = fftwf_plan_dft_c2r_1d(WAVETABLE_LENGTH , m_specBuf, m_sampleBuffer, FFTW_MEASURE);
+}
+
+void Oscillator::generateWaveTables()
+{
+	//generate sine tables
+	for (int i = 0; i < s_waveTablesPerWaveformCount; ++i)
+	{
+		generateSineWaveTable(MAX_FREQ / s_waveTableBandFreqs[i]);
+		s_waveTables[WaveShapes::SineWave][i] = m_generatedWaveTable;
+	}
+	//generate saw tables
+	for (int i = 0; i < s_waveTablesPerWaveformCount; ++i)
+	{
+		generateSawWaveTable(MAX_FREQ / s_waveTableBandFreqs[i]);
+		s_waveTables[WaveShapes::SawWave][i] = m_generatedWaveTable;
+	}
+	//generate square tables
+	for (int i = 0; i < s_waveTablesPerWaveformCount; ++i)
+	{
+		generateSquareWaveTable(MAX_FREQ / s_waveTableBandFreqs[i]);
+		s_waveTables[WaveShapes::SquareWave][i] = m_generatedWaveTable;
+	}
+	//generate triangle tables
+	for (int i = 0; i < s_waveTablesPerWaveformCount; ++i)
+	{
+		generateTriangleWaveTable(MAX_FREQ / s_waveTableBandFreqs[i]);
+		s_waveTables[WaveShapes::TriangleWave][i] = m_generatedWaveTable;
+	}
+	//generate moogSaw tables
+	//generate signal buffer
```

Comment:
I think it's possible to generate moog saw/exponential(actually parabolic) without FFT. Though their Fourier decompositions are not known well, but I can calculate it if you want. :wink:

Reply:
I was intending to keep the fft, As I am intending to use the same functionality for the user wave shapes, However, I would prefer to get the current scope implemented correctly first.

Reply:
The FFT analysis has remained, as the functionality has now been extended to the user waveforms